crates/modules/network-monitor/src/lib.rs

```diff
@@ -648,4 +664,83 @@ pub mod test_suite {
             )
             .report()
     }
+
+    fn dns_ipv4() -> TestCase {
+        TestCase::new(
+            "dns_ipv4",
+            run_dns(
+                SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::LOCALHOST, 18110)),
+                vec![IpAddr::V4(Ipv4Addr::LOCALHOST)],
+            ),
+        )
+    }
+
+    fn dns_ipv6() -> TestCase {
+        TestCase::new(
+            "dns_ipv6",
+            run_dns(
+                SocketAddr::V6(SocketAddrV6::new(Ipv6Addr::LOCALHOST, 18110, 0, 0)),
+                vec![IpAddr::V6(Ipv6Addr::LOCALHOST)],
+            ),
+        )
+    }
+
+    async fn run_dns(addr: SocketAddr, records: Vec<IpAddr>) -> TestReport {
+        TestRunner::with_ebpf(program)
+            .run_async(async move {
+                // DNS server.
+                let mut server = Server::default();
+                server.add_records("example.io", records).unwrap();
+                let socket = tokio::net::UdpSocket::bind(&addr).await.unwrap();
+                let local_addr = socket.local_addr().unwrap();
+                tokio::spawn(async move {
+                    server.start(socket).await.unwrap();
+                });
+
+                // DNS request.
+                let mut config = ResolverConfig::new();
+                config.add_name_server(NameServerConfig::new(local_addr, Protocol::Udp));
+                let resolver = TokioAsyncResolver::tokio(config, ResolverOpts::default());
+                resolver.lookup_ip("example.io").await.unwrap();
+            })
+            .await
+            .expect(move |event| {
+                if let NetworkEvent::Send { data, .. } = &event.payload {
+                    if data.is_empty() {
+                        return false;
+                    }
+                    let data = data.bytes(&event.buffer).unwrap();
+                    match parse_dns(data) {
+                        Some(Payload::DnsQuery { questions }) => {
+                            questions.iter().any(|q| q.name == "example.io")
+                        }
+                        _ => false,
+                    }
+                } else {
+                    false
+                }
+            })
+            .expect(move |event| {
+                if let NetworkEvent::Receive { data, .. } = &event.payload {
+                    if data.is_empty() {
+                        return false;
+                    }
+                    let data = data.bytes(&event.buffer).unwrap();
+                    match parse_dns(data) {
+                        Some(Payload::DnsResponse { questions, answers }) => {
+                            questions.iter().any(|q| q.name == "example.io")
+                                && answers.iter().any(|a| {
+                                    a.name == "example.io"
+                                        && (a.data == "A(Record(127.0.0.1))"
+                                            || a.data == "AAAA(Record(::1))")
```

Comment:
the record here is hardcoded, the test only works for localhost

Reply:
Well, I added the "example.io -> localhost" (both v4 and v6 representation of localhost) entry to the mock DNS and I'm asserting that exact event only. So I don't really see the problem.

Reply:
But I can add more domains and addresses if that's what you mean.

Reply:
Gotcha. Should be resolved now.